packages/peregrine/lib/talons/RootComponents/Product/useProduct.js

```diff
@@ -0,0 +1,111 @@
+import { useEffect, useMemo } from 'react';
+import { useApolloClient, useLazyQuery } from '@apollo/react-hooks';
+
+/**
+ * A [React Hook]{@link https://reactjs.org/docs/hooks-intro.html} that
+ * controls the logic for the Product Root Component.
+ *
+ * @kind function
+ *
+ * @param {object} props
+ *
+ */
+export const useProduct = props => {
+    const {
+        fragment,
+        query,
+        urlKey // TODO: move `getUrlKey` into Peregrine?
+    } = props;
+
+    const onServer = false;
+
+    const fragmentVariables = useMemo(() => {
+        return { onServer };
+    }, [onServer]);
+
+    const queryVariables = useMemo(() => {
+        return {
+            onServer,
+            urlKey
+        };
+    }, [onServer, urlKey]);
+
+    const [runQuery, { loading, error, data }] = useLazyQuery(query, {
+        // We're manually checking the cache here, so useLazyQuery can skip that.
+        fetchPolicy: 'network-only',
+        variables: queryVariables
+    });
+
+    const apolloClient = useApolloClient();
+
+    const productFromCache = useMemo(() => {
```

Comment:
After checking out the docs I'm not sure we need to do this fetch ourselves. Apollo has a [cache-and-network](https://www.apollographql.com/docs/react/api/react-apollo/) `fetch policy` which basically does the same thing that we have hand written here:

>This fetch policy will have Apollo first trying to read data from your cache. If all the data needed to fulfill your query is in the cache then that data will be returned. However, regardless of whether or not the full data is in your cache this `fetchPolicy` will always execute query with the network interface unlike `cache-first` which will only execute your query if the query data is not in your cache. This fetch policy optimizes for users getting a quick response while also trying to keep cached data consistent with your server data at the cost of extra network requests.

Can we just use the `cache-and-network`  policy in lieu of all this code?

Reply:
![image](https://user-images.githubusercontent.com/13182778/67576564-694b8e80-f704-11e9-8fc4-b92d6106454f.png)

DO YOU KNOW HOW LONG I LOOKED FOR THAT DOCUMENTATION?!!!!

I saw it as an option but I couldn't find exactly what it did - thank you!
I'll see if this works.

Reply:
Yeah, this should be the same as the service worker's [`stale-while-revalidate`](https://developers.google.com/web/fundamentals/instant-and-offline/offline-cookbook/#stale-while-revalidate).

Reply:
I should have realized this before, but no dice on this one. The whole reason we have to do this PR is because there's no single `product` query in Magento GraphQL - it will never find a cache hit unless we manually override it like this and tell it exactly where to look.

Reply:
I am simulating `cache-and-network` / `stale-while-revalidate` manually.

Reply:
Isn't `productDetails` filtered by `url_key` a "single product query"?

Reply:
To clarify, I mean a query that returns a `ProductInterface`. A `products` query (we rename it to `productDetails`) returns a `Products` object that nests `ProductInterface`s under `items`.

See https://github.com/magento/graphql-ce/issues/86#issuecomment-545935174 for a more detailed explanation.

I also used different words to describe it in the "does not use `cacheRedirects`" part of this PR description. 👍